File: src/main/java/org/prebid/cache/model/PayloadTransfer.java
package org.prebid.cache.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.Builder;
import lombok.RequiredArgsConstructor;
import lombok.Value;
import org.prebid.cache.helpers.Json;

@Value
@Builder(toBuilder = true)
@RequiredArgsConstructor
@JsonIgnoreProperties(ignoreUnknown = true)
public class PayloadTransfer {
    String type;
    String key;
    Object value;
    Long expiry;
    Long ttlseconds;
    transient String prefix;

    public String valueAsString() {
        if (value == null)
            return null;

        if (value instanceof String) {
            return value.toString();
        } else {
            return Json.toJson(value);
        }
    }

    public Long compareAndGetExpiry() {
        return ttlseconds != null ? ttlseconds : expiry;
    }
}

File: src/test/kotlin/org/prebid/cache/functional/model/request/PayloadTransfer.kt
package org.prebid.cache.functional.model.request

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL
import org.prebid.cache.functional.mapper.objectMapper
import org.prebid.cache.functional.model.request.MediaType.JSON
import org.prebid.cache.functional.model.request.MediaType.TEXT
import org.prebid.cache.functional.model.request.MediaType.XML
import org.prebid.cache.functional.util.getRandomString

@JsonInclude(NON_NULL)
data class PayloadTransfer(
    var type: MediaType,
    var value: String,
    var key: String? = null,
    var expiry: Long? = null,
    var ttlseconds: Long? = null,
    var prefix: String? = null,
    var application: String? = null,

    //Fields not affected on PBC
    var timestamp: Long? = null,
    var bidder: String? = null,
    var bidid: String? = null,
    var aid: String? = null,
) {

    companion object {
        private const val DEFAULT_EXPIRY = 300L
        private const val DEFAULT_TTL_SECONDS = 3000L

        private fun getDefaultPayloadTransfer(type: MediaType, value: String): PayloadTransfer =
            PayloadTransfer(type = type, value = value, expiry = DEFAULT_EXPIRY, ttlseconds = DEFAULT_TTL_SECONDS)

        fun getDefaultJsonPayloadTransfer(): PayloadTransfer =
            PayloadTransfer(
                type = JSON,
                value = objectMapper.writeValueAsString(TransferValue.getDefaultJsonValue()),
                expiry = 300
            )

        fun getDefaultXmlPayloadTransfer(): PayloadTransfer =
            getDefaultPayloadTransfer(XML, objectMapper.writeValueAsString(TransferValue.getDefaultXmlValue()))

        fun getDefaultTextPayloadTransfer(): PayloadTransfer =
            getDefaultPayloadTransfer(TEXT, getRandomString())
    }
}
